airflow/models/base_async_operator.py

```diff
@@ -0,0 +1,96 @@
+# -*- coding: utf-8 -*-
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+"""
+Base Asynchronous Operator for kicking off a long running
+operations and polling for completion with reschedule mode.
+"""
+from functools import wraps
+from airflow.sensors.base_sensor_operator import BaseSensorOperator
+from airflow.exceptions import AirflowException
+from airflow.models.xcom import XCOM_EXTERNAL_RESOURCE_ID_KEY
+
+class BaseAsyncOperator(BaseSensorOperator, SkipMixin):
```

Comment:
I think that this operator actually makes more sense as a parent of sensor operator, rather than as a subclass of sensor operator.  BaseSensorOperator is a BaseAsyncOperator that only checks status, does not submit request, and does not process result.

Or maybe you should actually just make these changes directly to BaseSensorOperator, making it more generic rather than subclassing it.  I am pretty sure you can do this while preserving backward compatibility.  Either they have common ancestor, or async is parent.

Either way, I think you need to orchestrate the submit request / process result from within the logic of sensor operator's current execute method.  That way you only call `submit_request` on first run (which in sensor operator would not do anything) and only process result after poke successful (which again would not do anything in the sensor operator case).

Reply:
did not see your changes to sensor

Reply:
I had a similar thought process. Where I ended up was the implementation of the common parent would nearly identical to `BaseSensorOperator`. The logic being `BaseAsyncOperator` just implements a few more methods (e.g `[submit/process]_request`) and is opinionated that `mode=reschedule`. 
The main reasoning to subclass v.s. enhancing the existing Sensor is I wanted to use the same `execute` method but have a class nomenclature that indicates that subclasses of this are intended to take action. I think people are less likely to understand that a Sensor can take action rather than being a read-only poller.

The spirit of this class is to provide an improved way for what we'd traditionally do with Operators so I wanted a class with the name Operator. Perhaps this is short sighted.

I've refactored to do the submit / process in the execute method of BaseSensorOperator

Reply:
Ref. the name, I agree that it's important for people to consider this as something that takes action and that's implied by the Operator suffix. That's why I consider BaseSensorOperator in its current form to have a misleading name: it implies that it's both for taking action and sensing. Luckily the actual derived sensor implementations are all suffixed with Sensor - e.g. EmrStepSensor - which doesn't mislead.

I feel like having "Async" in your new class highlights the wrong thing. It's not that it is async in the sense of non-synchronous or rescheduled (Airflow terminology) that makes this class unique, since all sensors derived from BaseSensorOperator *can* operate as async/rescheduled: it implies not occupying a task slot. What differentiates your class is that it combines action and sensing. That's why I tend to think that the already used BaseSensorOperator would actually be a great name to describe what your new class does; so why not enhance the existing class to optionally do the action part?

There might be technical reasons to have a separate class - e.g. code is easier to read or to reduce risk by extending rather than modifying (SOLID) - but I can't think of a better alternative to "Async": BaseDoThenWaitOperator (hah); BaseActionSensor; eventually back to BaseSensorOperator.